### PR TITLE
feat: update window.ai client library

### PR DESCRIPTION
**Pull request by louisgv**
pre-req: #30 <- I merged this into the PR as well to fix all the upcoming merging issue with package.json, also this PR merged main to align with latest. I would update `colin/window-lib` with main, which will make the diff easier to triage

This PR:
- Fix up client API
- Use the client API in both extension and web app
- Align all the typing with the client API's interface
- Removed the type package

### CODE REVIEW

File: turbo.json [hunk 1]
```diff
@@ -12,6 +12,7 @@
     },
     "test": {},
     "dev": {
+      "dependsOn": ["^build"],
```
Comment (alexanderatallah) [inline turbo.json:15]: Is this true? doesn't it not need to run the build step for the extension? (definitely not for web, right?)

File: packages/lib/src/index.ts [hunk 2]
```diff
@@ -42,17 +58,7 @@ export interface CompletionOptions {
...
-
-// ModelID is an enum of the available models.
-// NOTE: this is an evolving standard, and may change in the future.
-export enum ModelID {
-  GPT3 = "openai/gpt3.5",
-  GPT4 = "openai/gpt4",
-  GPTNeo = "together/gpt-neoxt-20B",
-  Cohere = "cohere/xlarge",
-  Local = "local"
+  model?: TModel
```
Comment (alexanderatallah) [inline packages/lib/src/index.ts:61]: good idea

File: packages/lib/src/index.ts [hunk 3]
```diff
@@ -72,36 +78,74 @@ export enum EventType {
...
+  event: EventType,
+  data: T | ErrorCode
+) => void
+
+export const VALID_DOMAIN = "https://windowai.io" as const
+
+export interface WindowAI<TModel = string> {
+  __window_ai_metadata__: {
+    domain: typeof VALID_DOMAIN
+    version: string
```
Comment (alexanderatallah) [inline packages/lib/src/index.ts:93]: 👍

File: packages/lib/src/index.ts [hunk 4]
```diff
@@ -72,36 +78,74 @@ export enum EventType {
...
-        if (window.ai) {
-          clearInterval(interval)
-          resolve(1)
-        }
-      }, 100)
-    })
+// Checking against other window.ai implementations
+export function hasWindowAI() {
+  return (
+    !!globalThis.window.ai?.__window_ai_metadata__ &&
```
Comment (alexanderatallah) [inline packages/lib/src/index.ts:115]: Unfortunately, we announced the extension and pointed people to download it prior to injecting this metadata into the window.ai object, so this is going to return false for all the prior users.

Open to other ideas here, but I think we should just check for the existence of `window.ai?.getCompletion` for now, and switch to this code down the road. (we could leave it commented here with a TODO)

File: packages/lib/src/index.ts [hunk 5]
```diff
@@ -72,36 +78,74 @@ export enum EventType {
...
+
+export type EventListenerHandler<T> = (
+  event: EventType,
+  data: T | ErrorCode
+) => void
+
+export const VALID_DOMAIN = "https://windowai.io" as const
+
+export interface WindowAI<TModel = string> {
+  __window_ai_metadata__: {
```
Comment (alexanderatallah) [inline packages/lib/src/index.ts:91]: Don't think we need to decide this quite yet, but just fyi, there's an alternative standard being discussed for crypto wallets here: https://eips.ethereum.org/EIPS/eip-5749

File: packages/lib/README.md [hunk 6]
```diff
@@ -1,16 +1,16 @@
 # Official window.ai client library.
 
-This is the official window.ai client library for Node.js. It provides convenient access to the window.ai API for client and server-side applications.
+This is the official window.ai client library. It provides the base `WindowAI` interface. The `window.ai` browser extension implements this interface to expose the window.ai API to web applications. Any web application can leverage this client library to consume the `window.ai` API.
```
Comment (alexanderatallah) [inline packages/lib/README.md:3]: I think we should add a note about how it's a convenience, e.g.:

> Note that you can interact with `window.ai` directly without dependencies, but this library will help you check for installation correctly and use correct types.

File: apps/extension/src/core/constants.ts [hunk 7]
```diff
@@ -13,31 +14,31 @@ export enum PortName {
...
+    id?: RequestID
+    request: { requesterId: RequestID; permitted?: boolean }
   }
-  [PortName.Model]: { id: RequestId; request: ModelRequest }
-  [PortName.Events]: { id?: RequestId; request: EventRequest<unknown> }
+  [PortName.Model]: { id: RequestID; request: ModelRequest }
+  [PortName.Events]: { id?: RequestID; request: EventRequest<unknown> }
 }
 
 export interface PortResponse {
```
Comment (alexanderatallah) [inline apps/extension/src/core/constants.ts:26]: (Not relevant to this PR, but would love to hear your thoughts on how to do port request/response types sometime. Sticking them all in one file and not the `ports` files seems bad.)

File: apps/web/components/Chat.tsx [hunk 8]
```diff
@@ -60,30 +69,44 @@ export function Chat() {
     }
   }, [cookie, setCookie])
 
+  // TODO: This efefct can be an util hook
+  useEffect(() => {
+    const init = async () => {
+      try {
+        // we can also just use the poolWindowAI method here, and use window.ai directly down there as well
```
Comment (alexanderatallah) [inline apps/web/components/Chat.tsx:76]: what do you mean by pool?

**Review by alexanderatallah — COMMENTED** [review]
Added comments - great work! excited to land this!

File: turbo.json (hunk 1 above)
Comment (alexanderatallah) [reply turbo.json:15]: oh wait nvm, the `^` thing

File: packages/lib/src/index.ts (hunk 3 above)
Comment (louisgv) [reply packages/lib/src/index.ts:93]: yeet

File: apps/web/components/Chat.tsx (hunk 8 above)
Comment (louisgv) [reply apps/web/components/Chat.tsx:76]: Argh I forgot to rename the function name in this comment. The client lib export the function `waitForWindowAI` which was originally named poolWindowAI

File: apps/extension/src/core/constants.ts (hunk 7 above)
Comment (louisgv) [reply apps/extension/src/core/constants.ts:26]: ref:

- https://github.com/PlasmoHQ/plasmo/issues/334
- https://docs.plasmo.com/framework/messaging#e2e-type-safety-wip

I'd say it depends - I found barrel patterns to be not flexible, but it's great to consolidate thought when planning out architecture. I generally do `import type {}` in the port consumer, and let the port files be the source of truth. The messaging/port API can incorporate better typing practice as well.

re: barrel https://flaming.codes/en/posts/barrel-files-in-javascript

File: packages/lib/README.md (hunk 6 above)
Comment (louisgv) [reply packages/lib/README.md:3]: Done 👍

**Review by alexanderatallah — APPROVED** [review]

**Pull request merged**